src/Microsoft.VisualStudio.Composition/Configuration/SerializationContextBase.cs

```diff
@@ -688,7 +702,16 @@ protected void Write(IReadOnlyDictionary<string, object> metadata)
 
                 // Special case certain values to avoid defeating lazy load later.
                 // Check out the ReadMetadata below, how it wraps the return value.
-                var serializedMetadata = new LazyMetadataWrapper(metadata.ToImmutableDictionary(), LazyMetadataWrapper.Direction.ToSubstitutedValue, this.Resolver);
+                IReadOnlyDictionary<string, object> serializedMetadata;
+                if (metadata is LazyMetadataWrapper)
+                {
+                    serializedMetadata = LazyMetadataWrapper.TryUnwrap(metadata);
```

Comment:
Yes, the issue is that if this is already a LazyMetaDataWrapper then we want the internal structure since it will contain representations of the metadata that don't require assembly loads. If its not a LazyMetadataWrapper then we need to wrap it to produce the serializable forms of the metadata. I'll add a comment explaining what's going on.

Reply:
That depends. The LazyMetadataWrapper goes both directions. Unwrapping it may actually expose the types and lead to you serializing those instead of TypeRefs.